File: leetcode/7-reverse-integer.py
# https://leetcode.com/problems/reverse-integer/ (top-interview-questions)

import math

class Solution:
    def reverse(self, x: int) -> int:
        MAX = 2**31 - 1 #  2147483647
        MIN = -2**31    # -2147483648

        reversed_x = 0

        while x:
            digit = int(math.fmod(x, 10))
            x = int(x/10)
            if ((reversed_x > MAX // 10 or (reversed_x == MAX // 10 and digit > MAX % 10)) or
                (reversed_x < MIN // 10 or (reversed_x == MIN // 10 and digit < MIN % 10))):
                return 0
            reversed_x = reversed_x * 10 + digit
        return reversed_x

File: leetcode/7-reverse-integer.js
// https://leetcode.com/problems/reverse-integer/ (top-interview-questions)

/**
 * @param {number} x
 * @return {number}
 */
var reverse = function(x) {
  const MAX = (2 ** 31) - 1;
  const MIN = -(2**31);

  let reversedX = 0;

  while (x !== 0) {
    const digit = x % 10;

    if ((reversedX > (MAX - digit) / 10) || (reversedX < (MIN - digit) / 10)) {
      /**
       * what's meaning reversedX > (MAX - digit) / 10 ?
       * => reversedX * 10 > (MAX - digit) => reversedX * 10 + digit > MAX
       * => so that's mean that when i will calculate the next value of reversedX
       * i will get a value greater then the MAX value ( 2^31 - 1 )
       * 
       * what's meaning reversedX < (MIN - digit) / 10 ?
       * => reversedX * 10 < (MIN - digit) => reversedX * 10 + digit < MIN
       * => so that's mean that when i will calculate the next value of reversedX
       * i will get a value less then the MIN value -(2^31)
       */
      return 0;
    }

    reversedX = reversedX * 10 + digit;
    x = Math.trunc(x / 10);

  }

  return reversedX;

};
